contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,156 @@
+
+## Design Details
+
+### `CSIDriver` Object
+
+#### Proposed API
+
+```go
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+//
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+    metav1.TypeMeta   `json:",inline"`
+
+    // Standard object metadata.
+    // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
+
+    // Specification describing the CSI volume driver and any custom
+    // configuration for it.
+    Spec CSIDriverSpec `json:"spec"`
+}
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+    // Standard list metadata
+    // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // Items is the list of CSIDrivers
+    Items []CSIDriver `json:"items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // Driver indicates the name of the CSI driver that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume driver requires an attach operation (because it
+    // implements the CSI ControllerPublishVolume() method), and that Kubernetes
+    // should call attach and wait for any attach operation to complete before
+    // proceeding to mounting.
+    // If value is not specified, default is true -- meaning attach will be
+    // called.
+    // +optional
+    AttachRequired *bool `json:"attachRequired"`
+    
+    // Indicates this CSI volume driver requires additional pod information
+    // (like podName, podUID, etc.) during mount operations.
+    // If this is set to true, Kubelet will pass pod information as
+    // VolumeAttributes in the CSI NodePublishVolume() calls.
+    // If value is not specified, default is false -- meaning pod information
+    // will not be passed on mount. 
+    // +optional
+    PodInfoRequiredOnMount *bool `json:"podInfoRequiredOnMount"`
+}
+
+```
+
+#### CRD Installation
+
+The `CSIDriver` object schema will checked in to kubernetes/kubernetes repo under a storage directory.
+The `CSIDriver` `CustomResourceDefinition` (CRD) will be installed by a new Kubernetes controller that is responsible for ensuring required CRDs are installed.
+The controller will periodically verify that the CRD is still installed, and recreate it, if it is not.
+
+#### CR Creation
+
+When a CSI volume driver is deployed on Kubernetes, it may optionally, register it self with Kubernetes by creating a new `CSIDriver` object.
+The Kubernetes team will modify the CSI [`driver-registrar` container](https://github.com/kubernetes-csi/driver-registrar) to automatically do this.
```

Comment:
hmm.... Thinking about this some more, with my operator hat on, I'd rather it not. If the driver-registrar were to do it, then we'd have to rbac it well in order to restrict its access to a safe subset of things. but if you are going through that much effort, you could instead just upload the crd directly in whatever csi packaging is provided, like via k8s yaml / helm chart to the cluster using your admin credentials instead and not worry about it at all. Similar to some of the reasoning Rook took when they tightened security (https://github.com/rook/rook/blob/master/design/security-model.md for details)